src/mlpack/methods/amf/init_rules/average_init.hpp

```diff
@@ -74,6 +72,57 @@ class AverageInitialization
     H = H + avgV;
   }
 
+   /**
+    * Initialize the matrix W or H to the average value of V with uniform
+    * random noise added.
+    *
+    * @param V Input matrix.
+    * @param r Rank of matrix.
+    * @param M W or H matrix, to be initialized to the average value of V
+    * with uniform random noise added.
+    * @param whichMatrix If true, initialize W. Otherwise, initialize H.
+    */
+  template<typename MatType>
+  inline static void InitializeOne(const MatType& V,
+                                   const size_t r,
+                                   arma::mat& M,
+                                   const bool whichMatrix = true)
+  {
+    const size_t n = V.n_rows;
+    const size_t m = V.n_cols;
+
+    double avgV = 0;
+    double min = DBL_MAX;
+
+    // Iterate over all elements in the matrix (for sparse matrices, this only
+    // iterates over nonzeros).
+    for (typename MatType::const_row_col_iterator it = V.begin();
+        it != V.end(); ++it)
+    {
+      avgV += *it;
+      // Track the minimum value.
+      if (*it < min)
+        min = *it;
+    }
+
+    avgV = sqrt(((avgV / (n * m)) - min) / r);
+
+    if (whichMatrix)
+    {
+      // Initialize W to random values
+      M.randu(n, r);
+
+      M = M + avgV;
+    }
+    else
+    {
+      // Initialize H to random values
+      M.randu(r, m);
+
+      M = M + avgV;
```

Comment:
`M += avgV`?  Or even... `M = arma::randu<arma::mat>(arma::size(M)) + avgV`?

Reply:
Just wanted to check, any reason to overlook this suggestion? I thought it would clean up the code a bit. Up to you.

Reply:
Oh, I've already changed it to `M += avgV` it in my branch. I thought "Resolve Conversation" just means I'll take this suggestion :)